### PR TITLE
Adds synchronous span APIs for datagram sockets.

### PR DESCRIPTION
**Pull request by PJB3005**
Hope I did this right.

Part of #33418/#938

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 1]
```diff
@@ -1380,6 +1380,67 @@ public int SendTo(byte[] buffer, EndPoint remoteEP)
             return SendTo(buffer, 0, buffer != null ? buffer.Length : 0, SocketFlags.None, remoteEP);
         }
 
+        /// <summary>
+        ///     Sends data to the specified endpoint.
```
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1384]: Nit:
```suggestion
        /// Sends data to the specified endpoint.
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 2]
```diff
@@ -1380,6 +1380,67 @@ public int SendTo(byte[] buffer, EndPoint remoteEP)
...
+        /// <exception cref="ArgumentNullException"><c>remoteEP</c> is <see langword="null" />.</exception>
+        /// <exception cref="SocketException">An error occurred when attempting to access the socket.</exception>
+        /// <exception cref="ObjectDisposedException">The <see cref="Socket"/> has been closed.</exception>
+        public int SendTo(ReadOnlySpan<byte> buffer, EndPoint remoteEP)
+        {
+            return SendTo(buffer, SocketFlags.None, remoteEP);
+        }
+
+        /// <summary>
+        ///     Sends data to a specific endpoint using the specified <see cref="SocketFlags"/>.
```
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1398]:
```suggestion
        /// Sends data to a specific endpoint using the specified <see cref="SocketFlags"/>.
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs [hunk 3]
```diff
@@ -300,31 +300,15 @@ public static unsafe SocketError SendFile(SafeSocketHandle handle, SafeFileHandl
             }
         }
 
-        public static unsafe SocketError SendTo(SafeSocketHandle handle, byte[] buffer, int offset, int size, SocketFlags socketFlags, byte[] peerAddress, int peerAddressSize, out int bytesTransferred)
+        public static SocketError SendTo(SafeSocketHandle handle, byte[] buffer, int offset, int size, SocketFlags socketFlags, byte[] peerAddress, int peerAddressSize, out int bytesTransferred) =>
+            SendTo(handle, new ReadOnlySpan<byte>(buffer, offset, size), socketFlags, peerAddress, peerAddressSize, out bytesTransferred);
```
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:304]:
```suggestion
            SendTo(handle, buffer.AsSpan(offset, size), socketFlags, peerAddress, peerAddressSize, out bytesTransferred);
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs [hunk 4]
```diff
@@ -300,31 +300,15 @@ public static unsafe SocketError SendFile(SafeSocketHandle handle, SafeFileHandl
...
+
+        public static unsafe SocketError SendTo(SafeSocketHandle handle, ReadOnlySpan<byte> buffer, SocketFlags socketFlags, byte[] peerAddress, int peerAddressSize, out int bytesTransferred)
         {
             int bytesSent;
-            if (buffer.Length == 0)
+            fixed (byte* bufferPtr = &MemoryMarshal.GetReference(buffer))
```
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:309]:
```suggestion
            fixed (byte* bufferPtr = buffer)
```
In the case `buffer` is empty this is guaranteed to return the null-ptr (to be precise: a Null-ref is returned which is then the null-ptr).

`MemoryMarshal.GetReference` doesn't provide that guarantee. So one should check for `IsEmpty` before. As it goes here to the kernel-call I'd go with the safe version.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs [hunk 5]
```diff
@@ -528,19 +512,16 @@ public static unsafe SocketError ReceiveMessageFrom(Socket socket, SafeSocketHan
             return SocketError.Success;
         }
 
-        public static unsafe SocketError ReceiveFrom(SafeSocketHandle handle, byte[] buffer, int offset, int size, SocketFlags socketFlags, byte[] socketAddress, ref int addressLength, out int bytesTransferred)
+        public static unsafe SocketError ReceiveFrom(SafeSocketHandle handle, byte[] buffer, int offset, int size, SocketFlags socketFlags, byte[] socketAddress, ref int addressLength, out int bytesTransferred) =>
+            ReceiveFrom(handle, new Span<byte>(buffer, offset, size), SocketFlags.None, socketAddress, ref addressLength, out bytesTransferred);
```
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:516]:
```suggestion
            ReceiveFrom(handle, buffer.AsSpan(offset, size), SocketFlags.None, socketAddress, ref addressLength, out bytesTransferred);
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs [hunk 6]
```diff
@@ -528,19 +512,16 @@ public static unsafe SocketError ReceiveMessageFrom(Socket socket, SafeSocketHan
...
+        public static unsafe SocketError ReceiveFrom(SafeSocketHandle handle, Span<byte> buffer, SocketFlags socketFlags, byte[] socketAddress, ref int addressLength, out int bytesTransferred)
         {
             int bytesReceived;
-            if (buffer.Length == 0)
-            {
-                bytesReceived = Interop.Winsock.recvfrom(handle, null, 0, socketFlags, socketAddress, ref addressLength);
-            }
-            else
+
+            fixed (byte* bufferPtr = &MemoryMarshal.GetReference(buffer))
```
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:522]:
```suggestion
            fixed (byte* bufferPtr = buffer)
```
See above.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs (hunk 4 above)
Comment (PJB3005) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:309]: The span versions of `Send` and `Receive` [do the same thing](https://github.com/dotnet/runtime/blob/2223babdd49118787c675e04aff711f936a10b26/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs#L278) which I copied. Should I change those too?

Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:309]: Had another thought about this. As you pass `buffer.Length` to the Win-API your code is OK, and please ignore my comment here.

FYI https://github.com/dotnet/runtime/issues/25138#issuecomment-375163012 handles that case is general.

**Comment by geoffkizer** [comment]
Thanks, this looks good to me.

@antonfirsov should review and approve. (I think he's out today but should be back.)

To be clear -- the ReceiveMessageFrom API is already done, right? I see it listed in #33418, but I think that's incorrect, right?

**Comment by PJB3005** [comment]
#46285 did `ReceiveMessageFrom`.

@antonfirsov is the way I did the tests correct? I didn't test manually (since I don't know how to compile against a local runtime like this).

Also the failing test seems to be a timeout in some HTTP API which I don't think I could have broken with these changes?

**Comment by geoffkizer** [comment]
> Also the failing test seems to be a timeout in some HTTP API which I don't think I could have broken with these changes?

Yeah, don't worry about the failing test. Test results look fine.

**Review by antonfirsov — COMMENTED** [review]
Product code looks good, although I have some questions to the team regarding `NetEventSource`.

Can we add 0-byte send tests similarly to #51473?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 7]
```diff
@@ -1380,6 +1380,67 @@ public int SendTo(byte[] buffer, EndPoint remoteEP)
...
+        public int SendTo(ReadOnlySpan<byte> buffer, SocketFlags socketFlags, EndPoint remoteEP)
+        {
+            ThrowIfDisposed();
+            if (remoteEP == null)
+            {
+                throw new ArgumentNullException(nameof(remoteEP));
+            }
+
+            ValidateBlockingMode();
+
```
[the remark this answers is not in the document]
Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1416]: NetEventSource is whatever we want for internal logging purposes. It's not intended to be public, ever. To the extent that it duplicates our public telemetry, we should remove it. But there's a lot of internal logging code that has proven useful in the past, and we don't want to just blindly rip this out.

[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1416]: OK, from what you said it feels like it's not worth opening an issue for it, but we should keep it in mind. A dedupe refactor across the span/array overloads should fix this anyways.

[the remark this answers is not in the document]
Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1416]: Yep.

> A dedupe refactor across the span/array overloads should fix this anyways.

This is blocked on #46600 right?

**Comment by geoffkizer** [comment]
What else needs to happen here before we merge? @antonfirsov?

**Comment by antonfirsov** [comment]
In my https://github.com/dotnet/runtime/pull/51956#pullrequestreview-648356203 I recommended to add tests for the 0-byte send tests similarly to #51473. Not essential, but would be nice to have.

**Comment by PJB3005** [comment]
Ah, my bad, didn't realize that was something I should do.

**Comment by PJB3005** [comment]
I added 0-byte send tests. Do these look good @antonfirsov?

**Review by antonfirsov — APPROVED** [review]
LGTM. Test failures are unrelated.

Thanks!

**Pull request merged**